File: python/problem_43.py
from typing import Optional


class MaxStack:
    def __init__(self):
        self._data = []
        self._maxs = []

    def push(self, val: int) -> None:
        self._data.append(val)
        if not self._maxs or val >= self._maxs[-1]:
            self._maxs.append(val)
        else:
            self._maxs.append(self._maxs[-1])

    def pop(self) -> Optional[int]:
        if not self._data:
            return None
        self._maxs.pop()
        return self._data.pop()

    def max(self) -> Optional[int]:
        if not self._maxs:
            return None
        return self._maxs[-1]


def main():
    print("Problem 43:")
    s = MaxStack()
    s.push(1); s.push(3); s.push(2)
    assert s.max() == 3
    assert s.pop() == 2
    assert s.max() == 3
    assert s.pop() == 3
    assert s.max() == 1
    assert MaxStack().pop() is None
    print("All Tests Passed")

File: rust/problem_43.rs
pub struct MaxStack {
    data: Vec<i64>,
    maxs: Vec<i64>,
}

impl MaxStack {
    pub fn new() -> Self { MaxStack { data: Vec::new(), maxs: Vec::new() } }

    pub fn push(&mut self, val: i64) {
        self.data.push(val);
        let m = self.maxs.last().copied().unwrap_or(i64::MIN).max(val);
        self.maxs.push(m);
    }

    pub fn pop(&mut self) -> Option<i64> {
        self.maxs.pop();
        self.data.pop()
    }

    pub fn max(&self) -> Option<i64> { self.maxs.last().copied() }
}

pub fn problem43() {
    println!("Problem 43:");
    let mut s = MaxStack::new();
    s.push(1); s.push(3); s.push(2);
    assert_eq!(s.max(), Some(3));
    assert_eq!(s.pop(), Some(2));
    assert_eq!(s.max(), Some(3));
    assert_eq!(s.pop(), Some(3));
    assert_eq!(s.max(), Some(1));
    assert_eq!(MaxStack::new().pop(), None);
    println!("All Tests Passed");
}
